#Script generated by: Aaron D. Ramirez-Sanchez 
#Goal: To do Differential Expression Analysis and create DESEQ object with model using age, batch; sex as interaction and get DEGs from dds object 

##ENVIRONMENT
library(DESeq2)
library(Matrix)
library(org.Hs.eg.db)
library(clusterProfiler)

##FUNCTIONS

#FUN: rownames2col, changes row names using ENSEMBL column
rownames2col <- function(inDF, RowName = "ENSEMBL") {
  temp <- data.frame(rownames(inDF), inDF, row.names = NULL)
  names(temp)[1] <- RowName
  temp
}

#FUN: ENSEMBLToEntrezSYMBOL, using ENSEMBL, gets Entrez and SYMBOL
ENSEMBLToEntrezSYMBOL <- function(f) {
  df <- f
  gene <- df$ENSEMBL
  gene.df <- bitr(gene, fromType = "ENSEMBL",
                  toType = c("ENTREZID", "SYMBOL"),
                  OrgDb = org.Hs.eg.db)
  df <- dplyr::full_join(df, gene.df, by = "ENSEMBL")
  df <- df[!duplicated(df$ENSEMBL),]
  return(df)
}

#FUN: add.down_or_up.column, to add extra column as downregulated or upregulated
add.down_or_up.column <- function(df){
  df$Direction <- "Upregulated"
  df$Direction[df$log2FoldChange < 0] <- "Downregulated"
  df$Direction <- as.factor(df$Direction)
  df$Direction <- factor(df$Direction, levels = c("Upregulated", "Downregulated"))
  return(df)
}

#FUN: sorting.L2Fc, to sort based on L2FC
sorting.L2Fc <- function(df){
  df <- df[order(df$log2FoldChange, decreasing = TRUE), ]
  return(df)
}

#FUN: trim, trim.3rd, to trim values based on IQ
trim <- function(x){
  x[(x > mean(x)-1.5*IQR(x)) & (x < mean(x)+1.5*IQR(x))]
}
trim.3rd <- function(x){
  x[(x < quantile(rowSums(counts(dds)), 0.75)) & (x >= 1)]
}


##INPUT

#Reading count matrix of gene expression
EL.rna.matrix <- as.matrix(read.delim(file = "./data/processed/gene_expression/RNA_matrix.table", sep="\t", row.names = 1))

#Reading metadata
coldata <- read.delim(file = "./data/processed/metadata/metadata.tsv", sep = "\t", row.names = "sample.new")


##MAIN

#Scaling and centering numeric variables to use in model
coldata$GC <- scale(coldata$GC, center = TRUE)
coldata$RIN.GS <- scale(coldata$RIN.GS, center = TRUE)
coldata$Total.reads <- scale(coldata$Total.reads, center = TRUE)
coldata$age2 <- scale(coldata$age2, center = TRUE)

#creating dds object using formula including relevant covariates
dds <- DESeqDataSetFromMatrix(countData = EL.rna.matrix,
                              colData = coldata,
                              design= ~ sequencing.batch + GC + RIN.GS + Total.reads + age2 + sex2 + condition)

#filtering based on author of deseq https://support.bioconductor.org/p/65091/
dds <- estimateSizeFactors(dds)
nc <- counts(dds, normalized=TRUE)
filter <- rowSums(nc >= 10) >= 10
ddsMF <- dds[filter,]

ddsMF <- DESeq(ddsMF)

#Getting DE genes

# calculate coefficient vectors for each group
mod_mat <- model.matrix(design(ddsMF), colData(ddsMF))
UCD <- colMeans(mod_mat[ddsMF$condition == "UCD", ])
TCD <- colMeans(mod_mat[ddsMF$condition == "TCD", ])
CTRL <- colMeans(mod_mat[ddsMF$condition == "CTRL", ])

# obtain results for each pairwise contrast
#UCDvsCTRL
df.UCDvsCTRL <- results(ddsMF, contrast = UCD - CTRL)
df.UCDvsCTRL <- df.UCDvsCTRL[order(df.UCDvsCTRL$padj), ]
df.UCDvsCTRL <- as.data.frame(df.UCDvsCTRL)
df.UCDvsCTRL <- rownames2col(df.UCDvsCTRL)
df.UCDvsCTRL <- ENSEMBLToEntrezSYMBOL(df.UCDvsCTRL)
#TCDvsCTRL
df.TCDvsCTRL <- results(ddsMF, contrast = TCD - CTRL)
df.TCDvsCTRL <- df.TCDvsCTRL[order(df.TCDvsCTRL$padj), ]
df.TCDvsCTRL <- as.data.frame(df.TCDvsCTRL)
df.TCDvsCTRL <- rownames2col(df.TCDvsCTRL)
df.TCDvsCTRL <- ENSEMBLToEntrezSYMBOL(df.TCDvsCTRL)
#UCDvsTCD
df.UCDvsTCD <- results(ddsMF, contrast = UCD - TCD)
df.UCDvsTCD <- df.UCDvsTCD[order(df.UCDvsTCD$padj), ]
df.UCDvsTCD <- as.data.frame(df.UCDvsTCD)
df.UCDvsTCD <- rownames2col(df.UCDvsTCD)
df.UCDvsTCD <- ENSEMBLToEntrezSYMBOL(df.UCDvsTCD)

#Merging DEGs

#adding direction
UCDvsCTRL <- add.down_or_up.column(df.UCDvsCTRL)
TCDvsCTRL <- add.down_or_up.column(df.TCDvsCTRL)
UCDvsTCD <- add.down_or_up.column(df.UCDvsTCD)
#adding extra column for comparison
UCDvsCTRL$Comparison <- "UCDvsCTRL"
TCDvsCTRL$Comparison <- "TCDvsCTRL"
UCDvsTCD$Comparison <- "UCDvsTCD"
#merging
DEG.merged <- rbind(UCDvsCTRL, TCDvsCTRL, UCDvsTCD)
DEG.merged <- sorting.L2Fc(DEG.merged)

##OUTPUT

#Saving dds object
saveRDS(ddsMF, "./output/objects/dds.RDS")

#Saving DE genes as csv
write.csv(df.UCDvsCTRL, 
          file="./output/tables/DEG/UCDvsCTRL.csv", 
          row.names = F, quote = F)
write.csv(df.TCDvsCTRL, 
          file="./output/tables/DEG/TCDvsCTRL.csv", 
          row.names = F, quote = F)
write.csv(df.UCDvsTCD, 
          file="./output/tables/DEG/UCDvsTCD.csv", 
          row.names = F, quote = F)

write.csv(DEG.merged, 
          file="./output/tables/DEG/comparisons_merged.csv", 
          row.names = F, quote = F)
