#Script generated by: Aaron D. Ramirez-Sanchez 
#Goal: To do Differential Expression Analysis and create DESEQ object with model using age, batch; sex as interaction and get DEGs from dds object 

##ENVIRONMENT
library(DESeq2)
library(Matrix)
library(org.Hs.eg.db)
library(clusterProfiler)

##FUNCTIONS

#FUN: rownames2col, changes row names using ENSEMBL column
rownames2col <- function(inDF, RowName = "ENSEMBL") {
  temp <- data.frame(rownames(inDF), inDF, row.names = NULL)
  names(temp)[1] <- RowName
  temp
}

#FUN: ENSEMBLToEntrezSYMBOL, using ENSEMBL, gets Entrez and SYMBOL
ENSEMBLToEntrezSYMBOL <- function(f) {
  df <- f
  gene <- df$ENSEMBL
  gene.df <- bitr(gene, fromType = "ENSEMBL",
                  toType = c("ENTREZID", "SYMBOL"),
                  OrgDb = org.Hs.eg.db)
  df <- dplyr::full_join(df, gene.df, by = "ENSEMBL")
  df <- df[!duplicated(df$ENSEMBL),]
  return(df)
}

#FUN: add.down_or_up.column, to add extra column as downregulated or upregulated
add.down_or_up.column <- function(df){
  df$Direction <- "Upregulated"
  df$Direction[df$log2FoldChange < 0] <- "Downregulated"
  df$Direction <- as.factor(df$Direction)
  df$Direction <- factor(df$Direction, levels = c("Upregulated", "Downregulated"))
  return(df)
}

#FUN: sorting.L2Fc, to sort based on L2FC
sorting.L2Fc <- function(df){
  df <- df[order(df$log2FoldChange, decreasing = TRUE), ]
  return(df)
}


##INPUT

#Reading count matrix of gene expression
dds <- readRDS("./output/objects/dds.RDS")

#Reading metadata
coldata <- read.delim(file = "./data/processed/metadata/extended_metadata.tsv", sep = "\t", row.names = 1)


##MAIN

colData(dds)$Cluster <-  coldata[,"Cluster"]
colData(dds)$condition.cluster <- paste0(colData(dds)$condition, "." , colData(dds)$Cluster)
colData(dds)$condition.cluster <- as.factor(colData(dds)$condition.cluster)

design(dds) <- formula(~ sequencing.batch + GC + RIN.GS + Total.reads + age2 + sex2 + condition.cluster)
dds <- DESeq(dds)

#Getting DE genes

# calculate coefficient vectors for each group
mod_mat <- model.matrix(design(dds), colData(dds))
UCD.Group_1 <- colMeans(mod_mat[dds$condition.cluster == "UCD.Group_1", ])
UCD.Group_2 <- colMeans(mod_mat[dds$condition.cluster == "UCD.Group_2", ])
UCD.Group_3 <- colMeans(mod_mat[dds$condition.cluster == "UCD.Group_3", ])

TCD.Group_1 <- colMeans(mod_mat[dds$condition.cluster == "TCD.Group_1", ])
TCD.Group_2 <- colMeans(mod_mat[dds$condition.cluster == "TCD.Group_2", ])
TCD.Group_3 <- colMeans(mod_mat[dds$condition.cluster == "TCD.Group_3", ])

CTRL.Group_1 <- colMeans(mod_mat[dds$condition.cluster == "CTRL.Group_1", ])
CTRL.Group_2 <- colMeans(mod_mat[dds$condition.cluster == "CTRL.Group_2", ])
CTRL.Group_3 <- colMeans(mod_mat[dds$condition.cluster == "CTRL.Group_3", ])

# obtain results for each pairwise contrast
#UCD.Group_3 vs UCD.Group_2
df.UCD.3vs2 <- results(dds, contrast = UCD.Group_3 - UCD.Group_2)
df.UCD.3vs2 <- df.UCD.3vs2[order(df.UCD.3vs2$padj), ]
df.UCD.3vs2 <- as.data.frame(df.UCD.3vs2)
df.UCD.3vs2 <- rownames2col(df.UCD.3vs2)
df.UCD.3vs2 <- ENSEMBLToEntrezSYMBOL(df.UCD.3vs2)
#TCD.Group_2 vs TCD.Group_1
df.TCD.2vs1 <- results(dds, contrast = TCD.Group_2 - TCD.Group_1)
df.TCD.2vs1 <- df.TCD.2vs1[order(df.TCD.2vs1$padj), ]
df.TCD.2vs1 <- as.data.frame(df.TCD.2vs1)
df.TCD.2vs1 <- rownames2col(df.TCD.2vs1)
df.TCD.2vs1 <- ENSEMBLToEntrezSYMBOL(df.TCD.2vs1)

#UCD.Group_2 vs TCD.Group_2
df.G2.UCDvsTCD <- results(dds, contrast = UCD.Group_2 - TCD.Group_2)
df.G2.UCDvsTCD <- df.G2.UCDvsTCD[order(df.G2.UCDvsTCD$padj), ]
df.G2.UCDvsTCD <- as.data.frame(df.G2.UCDvsTCD)
df.G2.UCDvsTCD <- rownames2col(df.G2.UCDvsTCD)
df.G2.UCDvsTCD <- ENSEMBLToEntrezSYMBOL(df.G2.UCDvsTCD)

#Merging DEGs

#adding direction
df.UCD.3vs2 <- add.down_or_up.column(df.UCD.3vs2)
df.TCD.2vs1 <- add.down_or_up.column(df.TCD.2vs1)
df.G2.UCDvsTCD <- add.down_or_up.column(df.G2.UCDvsTCD)

#adding extra column for comparison
df.UCD.3vs2$Comparison <- "UCD.3vs2"
df.TCD.2vs1$Comparison <- "TCD.2vs1"
df.G2.UCDvsTCD$Comparison <- "G2.UCDvsTCD"

#merging
DEG.merged <- rbind(df.UCD.3vs2, df.TCD.2vs1, df.G2.UCDvsTCD)
DEG.merged <- sorting.L2Fc(DEG.merged)

##OUTPUT

#Saving dds object
saveRDS(dds, "./output/objects/clusters_dds.RDS")

#Saving DE genes as csv
write.csv(df.UCD.3vs2, 
          file="./output/tables/DEG/UCD.3vs2.csv", 
          row.names = F, quote = F)
write.csv(df.TCD.2vs1, 
          file="./output/tables/DEG/TCD.2vs1.csv", 
          row.names = F, quote = F)
write.csv(df.G2.UCDvsTCD, 
          file="./output/tables/DEG/G2.UCDvsTCD.csv", 
          row.names = F, quote = F)

write.csv(DEG.merged, 
          file="./output/tables/DEG/clusters_comparisons_merged.csv", 
          row.names = F, quote = F)
