#Script generated by: Aaron D. Ramirez-Sanchez 
#Goal: To do Differential Expression Analysis and create DESEQ object with model using age, batch; sex as interaction and get DEGs from dds object 

##ENVIRONMENT
library(DESeq2)
library(Matrix)


##FUNCTIONS
correcting <- function(y, covs) {
  
  # Fit the linear model for each gene y to the the covariates  
  model   <- lm(y ~ ., data=covs, na.action=na.exclude, )
  
  # Extract the coefficients
  coef    <- model$coefficients
  
  # Calculate the value of x, after removing the variation that can be explained by the covariates 
  # e.g.  y = intercept + b1 * cov1 + b2 * cov2 + b3 * cov3 + residual error
  # We are interested in removing the part explained by only the covs which is:
  # [ b1 * cov1 + b2 * cov2 + b3 * cov3 ]
  # To get this part for each sample in y, we can use the predict(model) function, which will give:
  # intercept + b1 * cov1 + b2 * cov2 + b3 * cov3 
  # As we want to keep the intercept to keep the overall expression level, we remove this from the output
  # of predict(model) - coef[1]. the first item in the coefficients is the intercept.
  
  y.new <- y - (predict(model) - coef[1])
  
  return(y.new)
  
}

log.back.transforming <- function(m){
  m <- t(m)
  m <- (2^m)
  return(m)
}


##INPUT

#Reading dds object for log RNA matrix for decon eQTL
dds <- readRDS("./output/objects/dds.RDS")
#removing probes with zero variance
RNA.decon <- counts(dds)
RNA.decon <- RNA.decon[!rowVars(RNA.decon) == 0,]

#Reading metadata
coldata <- read.delim(file = "./data/processed/metadata/extended_metadata.tsv", sep = "\t", row.names = 1)

#Reading shared samples
shared.samples <- read.table(file = "./data/raw/metadata/shared.samplenames.old.txt")
shared.samples <- shared.samples$V1

##MAIN

#Filtering samples based on shared samples in FACS
coldata <- coldata[coldata$sample %in% shared.samples,]
RNA.bulk <- RNA.bulk[, rownames(coldata)]
RNA.decon <- RNA.decon[, rownames(coldata)]

#Scaling and centering numeric variables to use in model
coldata$sequencing.batch <- as.factor(coldata$sequencing.batch)
coldata$sex2 <- as.factor(coldata$sex2)
coldata$condition <- as.factor(coldata$condition)
coldata$marsh.2 <- as.factor(coldata$marsh.2)
coldata$GC <- scale(coldata$GC, center = TRUE)
coldata$RIN.GS <- scale(coldata$RIN.GS, center = TRUE)
coldata$Total.reads <- scale(coldata$Total.reads, center = TRUE)
coldata$age2 <- scale(coldata$age2, center = TRUE)
coldata$crypt_ratio <- scale(coldata$crypt_ratio, center = TRUE)

#Preparing covariates
covs.complete <- coldata
covs.marsh <- covs.complete[ ,c("marsh.2", "sequencing.batch", "age2", "sex2", "Total.reads", "GC", "RIN.GS")]
covs.crypt <- covs.complete[ ,c("crypt_ratio", "sequencing.batch", "age2", "sex2", "Total.reads", "GC", "RIN.GS")]
covs.condition <- covs.complete[ ,c("condition", "sequencing.batch", "age2", "sex2", "Total.reads", "GC", "RIN.GS")]
covs.mcc <- covs.complete[ ,c("marsh.2", "crypt_ratio", "condition", "sequencing.batch", "age2", "sex2", "Total.reads", "GC", "RIN.GS")]

#Preparing deconRNA
RNA.decon.log <- log2(RNA.decon+1)

#Correcting RNA matrix
corrected.marsh <- apply(RNA.decon.log, 1, correcting, covs = covs.marsh)
corrected.crypt <- apply(RNA.decon.log, 1, correcting, covs = covs.crypt)
corrected.condition <- apply(RNA.decon.log, 1, correcting, covs = covs.condition)
corrected.mcc <- apply(RNA.decon.log, 1, correcting, covs = covs.mcc)

#log back trnaformation
corrected.marsh <- log.back.transforming(corrected.marsh)
corrected.crypt <- log.back.transforming(corrected.crypt)
corrected.condition <- log.back.transforming(corrected.condition)
corrected.mcc <- log.back.transforming(corrected.mcc)

##OUTPUT

#Saving RNA matrix
write.table(corrected.marsh, 
            './output/tables/gene_expression/RNA_decon_marsh.table', 
            sep='\t',quote=F,col.names=NA)
write.table(corrected.crypt, 
            './output/tables/gene_expression/RNA_decon_crypt.table', 
            sep='\t',quote=F,col.names=NA)
write.table(corrected.condition, 
            './output/tables/gene_expression/RNA_decon_condition.table', 
            sep='\t',quote=F,col.names=NA)
write.table(corrected.mcc, 
            './output/tables/gene_expression/RNA_decon_mcc.table', 
            sep='\t',quote=F,col.names=NA)

write.table(rownames(coldata), './data/processed/metadata/shared.samplenames.new.txt',
            sep='\t',quote=F, row.names = F)
